File: day2/go/main.go
// Package main
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var colors = map[string]int{
	"red":   12,
	"green": 13,
	"blue":  14,
}

func main() {
	inputFile, err := os.Open("../input.txt")
	failOnError(err)
	defer inputFile.Close()

	var result int
	var result2 int
	scanner := bufio.NewScanner(inputFile)
	for scanner.Scan() {
		line := scanner.Text()
		first := strings.Split(line, ":")
		gameID := strings.Split(first[0], " ")[1]
		cubeSets := strings.Split(first[1], ";")
		possible := true
		var maxRed int
		var maxGreen int
		var maxBlue int
		for _, set := range cubeSets {
			for _, c := range strings.Split(set, ",") {
				cSplit := strings.Split(strings.TrimSpace(c), " ")
				count, err := strconv.Atoi(cSplit[0])
				if err != nil {
					fmt.Printf("Cannot convert to in: %s %s\n", cSplit[0], err.Error())
				}
				switch cSplit[1] {
				case "red":
					if maxRed < count {
						maxRed = count
					}
				case "green":
					if maxGreen < count {
						maxGreen = count
					}
				case "blue":
					if maxBlue < count {
						maxBlue = count
					}
				}
				if colors[cSplit[1]] < count && possible {
					possible = false
				}
			}
		}
		if possible {
			id, err := strconv.Atoi(gameID)
			if err != nil {
				fmt.Printf("Cannot convert to in: %s %s\n", gameID, err.Error())
			}
			result += id
		}
		result2 += maxRed * maxGreen * maxBlue
	}

	fmt.Printf("PART 1 RESULT: %d\n", result)
	fmt.Printf("PART 2 RESULT: %d\n", result2)
}

func failOnError(err error) {
	if err != nil {
		panic(err)
	}
}

File: day2/rust/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{prelude::*, BufReader, Result};

fn main() -> Result<()> {
    let _ = part_two();
    part_one()
}

fn part_one() -> Result<()> {
    let input_file = File::open("../../input.txt")?;
    let reader = BufReader::new(input_file);

    let colors: HashMap<_, _> = [("red", 12), ("green", 13), ("blue", 14)]
        .into_iter()
        .collect();

    let result = reader
        .lines()
        .enumerate()
        .filter_map(|(n, line)| {
            let line = line.unwrap();
            let mut iter = line.split_ascii_whitespace().skip(2);
            while let Some(x) = iter.next() {
                let count: u32 = x.parse().unwrap();
                let color = iter.next().unwrap().trim_end_matches([',', ';']);
                if count > colors[color] {
                    return None;
                }
            }
            Some(n + 1)
        })
        .sum::<usize>();

    println!("PART 1 RESULT: {}", result);

    Ok(())
}

fn part_two() -> Result<()> {
    let input_file = File::open("../../input.txt")?;
    let reader = BufReader::new(input_file);

    let result: u32 = reader
        .lines()
        .map(|line| {
            let mut colors: HashMap<String, u32> = [
                (String::from("red"), 0),
                (String::from("green"), 0),
                (String::from("blue"), 0),
            ]
            .into_iter()
            .collect();
            let line: &str = &line.unwrap();
            let mut iter = line.split_ascii_whitespace().skip(2);
            while let Some(x) = iter.next() {
                let count: u32 = x.parse().unwrap();
                let color = iter.next().unwrap().trim_end_matches([',', ';']);
                let max = colors.get_mut(color).unwrap();
                if count > *max {
                    *max = count;
                }
            }
            colors.into_iter().map(|(_, count)| count).product::<u32>()
        })
        .sum();

    println!("PART 2 RESULT: {}", result);

    Ok(())
}
